File: internal/integrations/confluence/actions.go
package confluence

import (
	"fmt"
	"regexp"

	"github.com/roee-hersh/hallpass/internal/catalog"
)

type resourceKind string

const (
	resPage     resourceKind = "page"
	resBlogpost resourceKind = "blogpost"
	resSpace    resourceKind = "space"
)

func (k resourceKind) idShape() string {
	if k == resSpace {
		return "KEY"
	}
	return "id"
}

// action is one Confluence operation. Page and blog post actions are asked
// of the content permission check with operation; space actions are matched
// against the space permission list by operation key and target type.
type action struct {
	name, desc string
	resource   resourceKind
	operation  string // content check operation, or space permission operation key
	target     string // space permission targetType (space actions only)
}

var actionList = []action{
	{name: "page.read", desc: "view a page", resource: resPage, operation: "read"},
	{name: "page.update", desc: "edit a page", resource: resPage, operation: "update"},
	{name: "page.delete", desc: "delete a page", resource: resPage, operation: "delete"},
	{name: "blogpost.read", desc: "view a blog post", resource: resBlogpost, operation: "read"},
	{name: "blogpost.update", desc: "edit a blog post", resource: resBlogpost, operation: "update"},
	{name: "blogpost.delete", desc: "delete a blog post", resource: resBlogpost, operation: "delete"},
	{name: "space.read", desc: "view the space (space permission read/space)", resource: resSpace, operation: "read", target: "space"},
	{name: "page.create", desc: "add pages in the space (create/page)", resource: resSpace, operation: "create", target: "page"},
	{name: "blogpost.create", desc: "add blog posts in the space (create/blogpost)", resource: resSpace, operation: "create", target: "blogpost"},
	{name: "comment.create", desc: "add comments in the space (create/comment)", resource: resSpace, operation: "create", target: "comment"},
	{name: "attachment.create", desc: "add attachments in the space (create/attachment)", resource: resSpace, operation: "create", target: "attachment"},
	{name: "space.export", desc: "export the space (export/space)", resource: resSpace, operation: "export", target: "space"},
	{name: "page.restrict", desc: "add or remove content restrictions in the space (restrict_content/space)", resource: resSpace, operation: "restrict_content", target: "space"},
	{name: "space.admin", desc: "administer the space (administer/space)", resource: resSpace, operation: "administer", target: "space"},
}

var actions = func() map[string]action {
	m := map[string]action{}
	for _, a := range actionList {
		m[a.name] = a
	}
	return m
}()

var (
	contentIDRe = regexp.MustCompile(`^[1-9][0-9]{0,18}$`)
	spaceKeyRe  = regexp.MustCompile(`^[A-Za-z0-9~_-]{1,255}$`)
)

// resource is a parsed page:<id>, blogpost:<id> or space:<KEY>.
type resource struct {
	kind resourceKind
	id   string
}

func parseResource(res catalog.Resource) (resource, error) {
	switch res.Type {
	case "page", "blogpost":
		if !contentIDRe.MatchString(res.ID) {
			return resource{}, fmt.Errorf("%s id %q must be a numeric content id", res.Type, res.ID)
		}
		return resource{kind: resourceKind(res.Type), id: res.ID}, nil
	case "space":
		if !spaceKeyRe.MatchString(res.ID) {
			return resource{}, fmt.Errorf("space key %q must match %s", res.ID, spaceKeyRe)
		}
		return resource{kind: resSpace, id: res.ID}, nil
	default:
		return resource{}, fmt.Errorf("resource type %q; use page:<id>, blogpost:<id> or space:<KEY>", res.Type)
	}
}

File: hallpass-py/src/hallpass/integrations/confluence/actions.py
"""The confluence actions and resource parsing."""

from __future__ import annotations

import re
from dataclasses import dataclass
from enum import Enum

from hallpass.core.catalog import Resource
from hallpass.core.errors import go_quote

__all__ = [
    "ACTIONS",
    "ACTION_LIST",
    "CONTENT_ID_RE",
    "SPACE_KEY_PATTERN",
    "SPACE_KEY_RE",
    "ConfluenceAction",
    "ConfluenceResource",
    "ResKind",
    "parse_resource",
]


class ResKind(str, Enum):
    PAGE = "page"
    BLOGPOST = "blogpost"
    SPACE = "space"

    def __str__(self) -> str:
        return self.value

    def id_shape(self) -> str:
        return "KEY" if self == ResKind.SPACE else "id"


@dataclass(frozen=True)
class ConfluenceAction:
    """One Confluence operation. Page and blog post actions are asked of the
    content permission check with operation; space actions are matched
    against the space permission list by operation key and target type."""

    name: str
    desc: str
    resource: ResKind
    # The content check operation, or the space permission operation key.
    operation: str
    # The space permission targetType (space actions only).
    target: str = ""


ACTION_LIST: tuple[ConfluenceAction, ...] = (
    ConfluenceAction("page.read", "view a page", ResKind.PAGE, "read"),
    ConfluenceAction("page.update", "edit a page", ResKind.PAGE, "update"),
    ConfluenceAction("page.delete", "delete a page", ResKind.PAGE, "delete"),
    ConfluenceAction("blogpost.read", "view a blog post", ResKind.BLOGPOST, "read"),
    ConfluenceAction("blogpost.update", "edit a blog post", ResKind.BLOGPOST, "update"),
    ConfluenceAction("blogpost.delete", "delete a blog post", ResKind.BLOGPOST, "delete"),
    ConfluenceAction("space.read", "view the space (space permission read/space)", ResKind.SPACE, "read", "space"),
    ConfluenceAction("page.create", "add pages in the space (create/page)", ResKind.SPACE, "create", "page"),
    ConfluenceAction("blogpost.create", "add blog posts in the space (create/blogpost)", ResKind.SPACE, "create", "blogpost"),
    ConfluenceAction("comment.create", "add comments in the space (create/comment)", ResKind.SPACE, "create", "comment"),
    ConfluenceAction("attachment.create", "add attachments in the space (create/attachment)", ResKind.SPACE, "create", "attachment"),
    ConfluenceAction("space.export", "export the space (export/space)", ResKind.SPACE, "export", "space"),
    ConfluenceAction("page.restrict", "add or remove content restrictions in the space (restrict_content/space)", ResKind.SPACE, "restrict_content", "space"),
    ConfluenceAction("space.admin", "administer the space (administer/space)", ResKind.SPACE, "administer", "space"),
)

ACTIONS: dict[str, ConfluenceAction] = {a.name: a for a in ACTION_LIST}

# Go: ^[1-9][0-9]{0,18}$ and ^[A-Za-z0-9~_-]{1,255}$, matched in full (Go's
# $ matches only at the end of the text; Python's also before a newline).
CONTENT_ID_RE = re.compile(r"[1-9][0-9]{0,18}")
SPACE_KEY_PATTERN = "^[A-Za-z0-9~_-]{1,255}$"
SPACE_KEY_RE = re.compile(r"[A-Za-z0-9~_-]{1,255}")


@dataclass(frozen=True)
class ConfluenceResource:
    """A parsed page:<id>, blogpost:<id> or space:<KEY>."""

    kind: ResKind
    id: str


def parse_resource(res: Resource) -> ConfluenceResource:
    """The resource as a Confluence target; ValueError with Go's message
    otherwise."""
    if res.type in ("page", "blogpost"):
        if not CONTENT_ID_RE.fullmatch(res.id):
            raise ValueError(f"{res.type} id {go_quote(res.id)} must be a numeric content id")
        return ConfluenceResource(ResKind(res.type), res.id)
    if res.type == "space":
        if not SPACE_KEY_RE.fullmatch(res.id):
            raise ValueError(f"space key {go_quote(res.id)} must match {SPACE_KEY_PATTERN}")
        return ConfluenceResource(ResKind.SPACE, res.id)
    raise ValueError(f"resource type {go_quote(res.type)}; use page:<id>, blogpost:<id> or space:<KEY>")
